README.md
[![npm version](https://badge.fury.io/js/react-block-ui.svg)](https://badge.fury.io/js/react-block-ui) [![Build Status](https://travis-ci.org/Availity/react-block-ui.svg?branch=master)](https://travis-ci.org/Availity/react-block-ui) [![Coverage Status](https://coveralls.io/repos/github/Availity/react-block-ui/badge.svg?branch=master)](https://coveralls.io/github/Availity/react-block-ui?branch=master)
# react-block-ui
Easy way to block the user from interacting with your UI.

## About
This library contains easy to use components to prevent the user from interacting with the user interface based on some conditions.

There is also a separate component which works with redux to help block and unblock automatically based on redux action types. See the ReduxBlockUi component for more details.

## Documentation
See: [https://availity.github.io/react-block-ui/](https://availity.github.io/react-block-ui/)

## Installation

Install `react-block-ui` via NPM

```sh
npm install --save react-block-ui react react-dom
```

Import the components you need, example:

```js
import BlockUi from 'react-block-ui';
import ReduxBlockUi from 'react-block-ui/redux'; // only for redux
import reduxMiddleware from 'react-block-ui/reduxMiddleware'; // only for redux
import from 'react-block-ui/style.css';
```

## Development

Install dependencies:

```sh
npm install
```

Run examples at [http://localhost:8080/](http://localhost:8080/) with webpack dev server:

```sh
npm start
```

Run tests:

```sh
npm test
```

Run tests & coverage report:

```sh
npm test:coverage
```

Watch tests:

```sh
npm run test:watch
```

## Disclaimer
Open source software components distributed or made available in the Availity Materials are licensed to Company under the terms of the applicable open source license agreements, which may be found in text files included in the Availity Materials.

## License
[MIT](./LICENSE)


CHANGELOG.md
<a name="1.3.6"></a>
## [1.3.6](https://github.com/Availity/react-block-ui/compare/1.3.5...1.3.6) (2022-12-22)



<a name="1.3.5"></a>
## [1.3.5](https://github.com/Availity/react-block-ui/compare/1.3.4...1.3.5) (2022-12-07)



<a name="1.3.4"></a>
## [1.3.4](https://github.com/Availity/react-block-ui/compare/1.3.3...1.3.4) (2022-12-06)


### Bug Fixes

* trigger keepInView when container is set ([#55](https://github.com/Availity/react-block-ui/issues/55)) ([6123a5f](https://github.com/Availity/react-block-ui/commit/6123a5f)), closes [#47](https://github.com/Availity/react-block-ui/issues/47)



<a name="1.3.3"></a>
## [1.3.3](https://github.com/Availity/react-block-ui/compare/1.3.2...1.3.3) (2020-03-11)



<a name="1.3.2"></a>
## [1.3.2](https://github.com/Availity/react-block-ui/compare/v1.3.1...v1.3.2) (2019-10-28)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/Availity/react-block-ui/compare/1.2.0...1.3.0) (2019-08-02)



<a name="1.2.0"></a>
# [1.2.0](https://github.com/Availity/react-block-ui/compare/1.1.4...1.2.0) (2019-08-02)


### Bug Fixes

* **Tag:** account for forwardRef objects ([863b7e4](https://github.com/Availity/react-block-ui/commit/863b7e4))



<a name="1.1.4"></a>
## [1.1.4](https://github.com/Availity/react-block-ui/compare/v1.1.3...v1.1.4) (2019-06-18)


### Bug Fixes

* **BlockUI:** allow tag to be components ([#27](https://github.com/Availity/react-block-ui/issues/27)) ([7b51125](https://github.com/Availity/react-block-ui/commit/7b51125)), closes [#8](https://github.com/Availity/react-block-ui/issues/8)
* **focus:** ensure elm can be focused/blurred ([#22](https://github.com/Availity/react-block-ui/issues/22)) ([4ad7107](https://github.com/Availity/react-block-ui/commit/4ad7107))



<a name="1.1.3"></a>
## [1.1.3](https://github.com/Availity/react-block-ui/compare/v1.1.2...v1.1.3) (2018-09-07)


### Bug Fixes

* **FocusManagement:** fix scroll to top on unlock ([#20](https://github.com/Availity/react-block-ui/issues/20)) ([58c9e67](https://github.com/Availity/react-block-ui/commit/58c9e67))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/Availity/react-block-ui/compare/v1.1.1...v1.1.2) (2018-07-31)


### Bug Fixes

* **activeElement:** account for IE suckage ([7628259](https://github.com/Availity/react-block-ui/commit/7628259)), closes [#18](https://github.com/Availity/react-block-ui/issues/18)



<a name="1.1.1"></a>
## [1.1.1](https://github.com/Availity/react-block-ui/compare/v1.1.0...v1.1.1) (2017-11-01)


### Bug Fixes

* **styles:** fix bad comment syntax. ([4b4fc68](https://github.com/Availity/react-block-ui/commit/4b4fc68))

### Features


* **keepInView:** add keepInView prop ([#15](https://github.com/Availity/react-block-ui/pull/15)) ([7ba8b8e](https://github.com/Availity/react-block-ui/commit/7ba8b8e6)), closes [#14](https://github.com/Availity/react-block-ui/issues/14)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/Availity/react-block-ui/compare/v1.0.2...v1.0.3) (2017-09-07)


### Bug Fixes

* **focus:** fallback to setTimeout ([cd5d8c4](https://github.com/Availity/react-block-ui/commit/cd5d8c4)), closes [#11](https://github.com/Availity/react-block-ui/issues/11)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/Availity/react-block-ui/compare/v1.0.1...v1.0.2) (2017-08-04)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/Availity/react-block-ui/compare/v1.0.0...v1.0.1) (2017-08-04)


### Bug Fixes

* **docs:** fix generating docs for gh-pages ([687f454](https://github.com/Availity/react-block-ui/commit/687f454))
* **npm:** fix npm release ([ab4ac0b](https://github.com/Availity/react-block-ui/commit/ab4ac0b))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/Availity/react-block-ui/compare/0.2.6...v1.0.0) (2017-08-03)



<a name="0.2.6"></a>
## [0.2.6](https://github.com/Availity/react-block-ui/compare/0.2.5...0.2.6) (2017-04-17)



<a name="0.2.5"></a>
## [0.2.5](https://github.com/Availity/react-block-ui/compare/0.2.4...0.2.5) (2017-04-04)


### Bug Fixes

* **Focus:** add helper element to access wrapping element ([d62652f](https://github.com/Availity/react-block-ui/commit/d62652f)), closes [#8](https://github.com/Availity/react-block-ui/issues/8)



<a name="0.2.4"></a>
## [0.2.4](https://github.com/Availity/react-block-ui/compare/0.2.3...0.2.4) (2017-02-01)


### Bug Fixes

* **focus:** add null check for topFocus ([fb1bc3e](https://github.com/Availity/react-block-ui/commit/fb1bc3e)), closes [#5](https://github.com/Availity/react-block-ui/issues/5)



<a name="0.2.3"></a>
## [0.2.3](https://github.com/Availity/react-block-ui/compare/0.2.2...0.2.3) (2017-01-30)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/Availity/react-block-ui/compare/0.2.1...0.2.2) (2016-11-29)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/Availity/react-block-ui/compare/0.2.0...0.2.1) (2016-11-22)


### Bug Fixes

* **deps:** fix broken devDep ([20e52c5](https://github.com/Availity/react-block-ui/commit/20e52c5))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/Availity/react-block-ui/compare/0.1.1...0.2.0) (2016-11-22)


### Features

* **loader:** add ability to specify custom loading content ([09de8bf](https://github.com/Availity/react-block-ui/commit/09de8bf))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/Availity/react-block-ui/compare/0.1.0...0.1.1) (2016-09-07)


### Bug Fixes

* **package:** include css and redux at root ([e70559b](https://github.com/Availity/react-block-ui/commit/e70559b))



<a name="0.1.0"></a>
# 0.1.0 (2016-09-07)





src/BlockUi.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import DefaultLoader from './Loader';
import safeActiveElement from './safeActiveElement';

const defaultProps = {
  tag: 'div',
  renderChildren: true,
  loader: DefaultLoader,
};

class BlockUi extends Component {
  constructor(props) {
    super(props);

    this.tabbedUpTop = this.tabbedUpTop.bind(this);
    this.tabbedDownTop = this.tabbedDownTop.bind(this);
    this.tabbedUpBottom = this.tabbedUpBottom.bind(this);
    this.tabbedDownBottom = this.tabbedDownBottom.bind(this);
    this.setHelper = this.setRef.bind(this, 'helper');
    this.setBlocker = this.setRef.bind(this, 'blocker');
    this.setTopFocus = this.setRef.bind(this, 'topFocus');
    this.setContainer = this.setRef.bind(this, 'container');
    this.setMessageContainer = this.setRef.bind(this, 'messageContainer');
    this.handleScroll = this.handleScroll.bind(this);

    this.state = { top: '50%' };
  }


  componentWillReceiveProps(nextProps) {
    if (nextProps.blocking !== this.props.blocking) {
      if (nextProps.blocking) {
        // blocking started
        if (this.helper && this.helper.parentNode && this.helper.parentNode.contains
          && this.helper.parentNode.contains(safeActiveElement())) {
          this.focused = safeActiveElement();
          // https://www.tjvantoll.com/2013/08/30/bugs-with-document-activeelement-in-internet-explorer/#blurring-the-body-switches-windows-in-ie9-and-ie10
          if (this.focused && this.focused !== document.body) {
            (window.setImmediate || setTimeout)(() => this.focused && typeof this.focused.blur === 'function' && this.focused.blur());
          }
        }
      } else {
        this.detachListeners();
        const ae = safeActiveElement();
        if (this.focused && (!ae || ae === document.body || ae === this.topFocus)) {
          if (typeof this.focused.focus === 'function') {
            this.focused.focus();
          }
          this.focused = null;
        }
      }
    }
    if (nextProps.keepInView && (nextProps.keepInView !== this.props.keepInView || (nextProps.blocking && nextProps.blocking !== this.props.blocking))) {
      this.attachListeners();
      this.keepInView(nextProps);
    }
  }

  componentWillUnmount() {
    this.detachListeners();
  }

  setRef(name, ref) {
    this[name] = ref;
    if (ref && name === 'container') {
      this.keepInView();
    }
  }

  attachListeners() {
    window.addEventListener('scroll', this.handleScroll);
  }

  detachListeners() {
    window.removeEventListener('scroll', this.handleScroll);
  }

  blockingTab(e, withShift = false) {
    // eslint-disable-next-line eqeqeq
    return this.props.blocking && (e.key === 'Tab' || e.keyCode === 9) && e.shiftKey == withShift;
  }

  tabbedUpTop(e) {
    if (this.blockingTab(e)) {
      this.blocker.focus();
    }
  }

  tabbedDownTop(e) {
    if (this.blockingTab(e)) {
      e.preventDefault();
      this.blocker.focus();
    }
  }

  tabbedUpBottom(e) {
    if (this.blockingTab(e, true)) {
      this.topFocus.focus();
    }
  }

  tabbedDownBottom(e) {
    if (this.blockingTab(e, true)) {
      e.preventDefault();
      this.topFocus.focus();
    }
  }

  keepInView(props = this.props) {
    if (props.blocking && props.keepInView && this.container) {
      const containerBounds = this.container.getBoundingClientRect();
      const windowHeight = window.innerHeight;
      if (containerBounds.top > windowHeight || containerBounds.bottom < 0) return;
      if (containerBounds.top >= 0 && containerBounds.bottom <= windowHeight) {
        if (this.state.top !== '50%') {
          this.setState({ top: '50%' });
        }
        return;
      }

      const messageBoundsHeight = this.messageContainer ? this.messageContainer.getBoundingClientRect().height : 0;
      let top = Math.max(Math.min(windowHeight, containerBounds.bottom) - Math.max(containerBounds.top, 0), messageBoundsHeight) / 2;
      if (containerBounds.top < 0) {
        top = Math.min(top - containerBounds.top, containerBounds.height - (messageBoundsHeight / 2));
      }
      if (this.state.top !== top) {
        this.setState({top});
      }

    }
  }

  handleScroll() {
    this.keepInView();
  }

  render() {
    const {
      tag: Tag,
      blocking,
      className,
      children,
      message,
      loader: Loader,
      renderChildren,
      keepInView,
      ariaLabel = 'loading',
      ...attributes
    } = this.props;

    const classes = blocking ? `block-ui ${className}` : className;
    const renderChilds = !blocking || renderChildren;

    return (
      <Tag {...attributes} className={classes} aria-busy={blocking}>
        {blocking &&
        <div tabIndex="0" onKeyUp={this.tabbedUpTop} onKeyDown={this.tabbedDownTop} ref={this.setTopFocus}>
          <div className="sr-only">
            {message || ariaLabel}
          </div>
        </div>}
        {renderChilds && children}
        {blocking &&
        <div className="block-ui-container"
          tabIndex="0"
          ref={this.setBlocker}
          onKeyUp={this.tabbedUpBottom}
          onKeyDown={this.tabbedDownBottom}
        >
          <div className="block-ui-overlay" ref={this.setContainer} />
          <div className="block-ui-message-container"
            ref={this.setMessageContainer}
            style={{ top: keepInView ? this.state.top : undefined }}
          >
            <div className="block-ui-message">
              {message || <span className="sr-only">{ariaLabel}</span>}
              <div aria-hidden>
                {React.isValidElement(Loader) ? Loader : <Loader/>}
              </div>
            </div>
          </div>
        </div>
        }
        <span ref={this.setHelper} />
      </Tag>
    );
  }
}

BlockUi.propTypes = {
  blocking: PropTypes.bool,
  children: PropTypes.node,
  renderChildren: PropTypes.bool,
  keepInView: PropTypes.bool,
  className: PropTypes.string,
  message: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  loader: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.node,
  ]),
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.shape({ $$typeof: PropTypes.symbol, render: PropTypes.func }),
  ]),
  ariaLabel: PropTypes.string,
};
BlockUi.defaultProps = defaultProps;

export default BlockUi;
